concordium-consensus/src/Concordium/GlobalState/Persistent/BlockState.hs

```diff
@@ -1747,220 +2035,496 @@ doConstrainBakerCommission pbs ai ranges = do
-    checkOverdelegation oldCapital oldTarget cp = do
-        let doCheckOverDelegation = do
-                let pp = cp ^. cpPoolParameters
-                ad <- getAccountOrFail
-                let target = ad ^. BaseAccounts.delegationTarget
-                bsp <- MTL.get
-                delegationConfigureDisallowOverdelegation bsp pp target
-        case (dcuCapital, dcuDelegationTarget) of
-            (Just newCapital, Just newTarget) -> unless (newCapital <= oldCapital && newTarget == oldTarget) doCheckOverDelegation
-            (Just newCapital, Nothing) -> unless (newCapital <= oldCapital) doCheckOverDelegation
-            (Nothing, Just newTarget) -> unless (newTarget == oldTarget) doCheckOverDelegation
-            _ -> return ()
+                        let delta = capital - oldCapital
+                        bsp1 <-
+                            onActiveBakers bsp $
+                                fmap (totalActiveCapital %~ addActiveCapital delta)
+                                    . modifyPoolCapitalUnsafe target (+ delta)
+                        maybeCooldownsBefore <- accountCooldowns acc
+                        acc1 <-
+                            setAccountStake capital acc
+                                >>= reactivateCooldownAmount delta
+                        maybeCooldownsAfter <- accountCooldowns acc1
+                        let removals = cooldownRemovals maybeCooldownsBefore maybeCooldownsAfter
+                        newCooldowns <- applyCooldownRemovalsGlobally ai removals (bspAccountsInCooldown bsp1)
+                        return (bsp1{bspAccountsInCooldown = newCooldowns}, acc1)
+    onActiveBakers bsp f = do
+        newPABRef <- refMake =<< f =<< refLoad (bspBirkParameters bsp ^. birkActiveBakers)
+        return bsp{bspBirkParameters = bspBirkParameters bsp & birkActiveBakers .~ newPABRef}
+    addToPrePreCooldowns ::
+        (MonadBlobStore m', PVSupportsFlexibleCooldown pv) =>
+        BlockStatePointers pv ->
+        m' (BlockStatePointers pv)
+    addToPrePreCooldowns bsp = do
+        -- Add the account to the pre-pre-cooldowns list.
+        newAccountsInCooldown <-
+            (accountsInCooldown . prePreCooldown)
+                (consAccountList ai)
+                (bspAccountsInCooldown bsp)
+        return bsp{bspAccountsInCooldown = newAccountsInCooldown}
+
+data CooldownRemovals = CooldownRemovals
+    { -- | Whether the pre-pre cooldown was removed.
+      crPrePreCooldown :: Bool,
+      -- | Whether the pre cooldown was removed.
+      crPreCooldown :: Bool,
+      -- | If all cooldowns were removed, this is the previous timestamp of the earliest cooldown.
+      crCooldown :: Maybe Timestamp
+    }
+
+-- | Determine if a change in cooldowns requires global updates to the indexes.
+--  The change should arise from (possibly) reactivating stake from cooldown.
+cooldownRemovals ::
+    Maybe CooldownQueue.Cooldowns -> Maybe CooldownQueue.Cooldowns -> CooldownRemovals
```

Comment:
Could you comment the inputs? What's the first/second cooldown queue?

Reply:
Done.